### PR TITLE
Commit implementing multiple features

### PR DESCRIPTION
**Pull request by misrasaurabh1**
* Implemented early stopping
* Implement tensorboard logging
* Multi-GPU secondary training and fixes
* Create a wrapper script for automating training and guessing
* Checkin code for graphing guesses
* Add code to convert enum_ofile to a file format for graphing

### CODE REVIEW

File: pwd_guess.py [hunk 1]
```diff
@@ -930,11 +950,16 @@ def train_model(self, serializer):
         max_accuracy = 0
         if self.config.scheduled_sampling:
             self.ctable.init_model(self.model)
+            if self.config.tensorboard:
+            self.callback.set_model(self.model)
```
Comment (wrmelicher) [inline pwd_guess.py]: This line looks suspicious. Did you mean to indent it so that it is guarded by the if statement? 

BTW, the pylint tool will catch this sort of thing if you run it before commits. Also the unittests fail because of it.

File: pwd_guess.py [hunk 2]
```diff
@@ -1007,6 +1063,15 @@ def retrain_classification(self, preprocessor, serializer):
...
         self.train(serializer)
 
+    def write_log(self, names, logs, batch_no):
+        assert(self.callback)
+        for name, value in zip(names, logs):
+            summary = tf.Summary()
+            summary_value = summary.value.add()
+            summary_value.simple_value = value
+            summary_value.tag = name
+            self.callback.writer.add_summary(summary, batch_no)
```
Comment (wrmelicher) [inline pwd_guess.py]: Unit tests seem to fail on this line.

File: pwd_guess.py [hunk 3]
```diff
@@ -826,6 +830,13 @@ def __init__(self, pwd_list, config=ModelDefaults(), multi_gpu=1):
         self.model_to_save = None
         self.multi_gpu = multi_gpu
         self.pwd_list = pwd_list
+        self.cumulative_chunks = 0
+        self.min_loss_early_stopping = 999999999 #A really large number
+        self.poor_batches_early_stopping = 0
+        if config.tensorboard:
+            self.callback = TensorBoard(".")
```
Comment (wrmelicher) [inline pwd_guess.py]: Can we make the path to put the tensorboard logs configurable either with a Config parameter or a command line argument?

File: pwd_guess.py [hunk 4]
```diff
@@ -580,6 +583,7 @@ class ModelDefaults(object):
     dropouts = False
     dropout_ratio = .25
     fuzzy_training_smoothing = False
+    tensorboard = True
```
Comment (wrmelicher) [inline pwd_guess.py]: Does it make sense to default to False here? I would prefer to preserve the typical behavior for tests so that the unit tests don't make non-temporary files.

File: pwd_guess.py [hunk 5]
```diff
@@ -965,31 +990,62 @@ def training_step(self, x_all, y_all, w_all):
             x_val, y_val, sample_weight=w_val)
         return (train_loss, train_accuracy, test_loss, test_accuracy)
 
-    def train_model_generation(self):
+    def early_stopping(self,smooth_loss, serializer):
```
Comment (wrmelicher) [inline pwd_guess.py]: Please use the pylint style here. To run it on this file, you can do "pylint --rcfile=pylintrc pwd_guess.py" There are a handful of other style comments that the pylint picks up too.

File: pwd_guess.py [hunk 6]
```diff
@@ -826,6 +830,13 @@ def __init__(self, pwd_list, config=ModelDefaults(), multi_gpu=1):
         self.model_to_save = None
         self.multi_gpu = multi_gpu
         self.pwd_list = pwd_list
+        self.cumulative_chunks = 0
+        self.min_loss_early_stopping = 999999999 #A really large number
```
Comment (wrmelicher) [inline pwd_guess.py]: Better to use float('inf') here instead just in case the loss is above 999..99.

File: pwd_guess.py [hunk 7]
```diff
@@ -826,6 +830,13 @@ def __init__(self, pwd_list, config=ModelDefaults(), multi_gpu=1):
...
+        self.poor_batches_early_stopping = 0
+        if config.tensorboard:
+            self.callback = TensorBoard(".")
+            self.train_log_names = []
+            self.test_log_names = []
```
Comment (wrmelicher) [inline pwd_guess.py]: It would be good to initialize these to None in the case where tensorboard is false so that we don't get confusing-to-debug attribute undefined errors later.

File: pwd_guess.py (hunk 1 above)
Comment (wrmelicher) [reply pwd_guess.py]: Also, should probably in general move the set_model call so that it is outside the check for scheduled_sampling.

File: pwd_guess.py [hunk 8]
```diff
@@ -965,31 +990,62 @@ def training_step(self, x_all, y_all, w_all):
...
+            tr_loss, tr_acc, te_loss, te_acc = self.training_step(
                 x_all, y_all, w_all)
             accuracy_accum += [(len(x_all), te_acc)]
+            smoothened_loss.append((len(x_all),te_loss))
+            if self.config.tensorboard:
+                self.write_log(self.train_log_names,[tr_loss, tr_acc], self.cumulative_chunks)
+                self.write_log(self.test_log_names,[te_loss,te_acc], self.cumulative_chunks)
             if chunk % self.config.chunk_print_interval == 0:
+                instances_smoothened = map(lambda x: x[0], smoothened_loss)
+                loss_smoothened = sum(map(lambda x: x[0] * x[1], smoothened_loss)) / sum(instances_smoothened)
```
Comment (wrmelicher) [inline pwd_guess.py]: Why are you multiplying the loss times the number of training instances here. I would have thought that the loss was already aggregated across all training instances. Is that wrong? If you think this is correct, maybe add a comment explaining why multiplying here is best. Also, using a named tuple here would be more readable instead of having to figure out what the 0 and 1 indexes are.

File: pwd_guess.py [hunk 9]
```diff
@@ -965,31 +990,62 @@ def training_step(self, x_all, y_all, w_all):
...
-                logging.info('Train loss %s. Test loss %s. Test accuracy %s.',
-                             tr_loss, te_loss, te_acc)
+                logging.info('Train loss %s. Test loss %s. Test accuracy %s. Averaged loss %s',
+                             tr_loss, te_loss, te_acc, loss_smoothened)
+            if self.config.early_stopping and self.cumulative_chunks >= self.config.early_stopping_patience and \
+                self.cumulative_chunks % self.config.chunk_print_interval == 0:
+                #Second condition so that the model doesn't start saving very early in the training process
+                #Third condition to prevent evaluation of accuracy too frequently
+                instances_smoothened = map(lambda x: x[0], smoothened_loss)
+                loss_smoothened = sum(map(lambda x: x[0] * x[1], smoothened_loss)) / sum(instances_smoothened)
```
Comment (wrmelicher) [inline pwd_guess.py]: This is the same calculation as above. Could they be merged so there is less repetition?

File: pwd_guess.py [hunk 10]
```diff
@@ -826,6 +830,13 @@ def __init__(self, pwd_list, config=ModelDefaults(), multi_gpu=1):
         self.model_to_save = None
         self.multi_gpu = multi_gpu
         self.pwd_list = pwd_list
+        self.cumulative_chunks = 0
```
Comment (wrmelicher) [inline pwd_guess.py]: This variable only seems to be used inside  of train_model_generation.  Could it be moved as a local variable in that function?

File: pwd_guess.py [hunk 11]
```diff
@@ -1007,6 +1063,15 @@ def retrain_classification(self, preprocessor, serializer):
...
 
+    def write_log(self, names, logs, batch_no):
+        assert(self.callback)
+        for name, value in zip(names, logs):
+            summary = tf.Summary()
+            summary_value = summary.value.add()
+            summary_value.simple_value = value
+            summary_value.tag = name
+            self.callback.writer.add_summary(summary, batch_no)
+            self.callback.writer.flush()
```
Comment (wrmelicher) [inline pwd_guess.py]: Could we only flush the logs on the chunk_print_interval so that we avoid flushing every iteration (and spending the time to do it)?

File: pwd_guess_unit.py [hunk 12]
```diff
@@ -395,7 +395,7 @@ def test_accuracy(self):
         mock_model.train_on_batch = MagicMock(return_value = (0.5, 0.5))
         mock_model.test_on_batch = MagicMock(return_value = (0.5, 0.5))
         t.model = mock_model
-        self.assertEqual(0.5, t.train_model_generation())
+        self.assertEqual((0.5, False), t.train_model_generation())
```
Comment (wrmelicher) [inline pwd_guess_unit.py]: It would be good to add some tests for the early stopping behavior. I could imagine a test that checks that the state in the early_stopping function works.

Also, if tensorboad is turned off by default, then adding a test with it on would be good.

File: pwd_guess.py (hunk 10 above)
Comment (misrasaurabh1) [reply pwd_guess.py]: This function is aggregating chunk counts across different function counts so it can't be a local variable within that function

Comment (wrmelicher) [reply pwd_guess.py]: OK, that's fine then.
